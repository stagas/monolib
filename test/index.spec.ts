require('fast-text-encoding')

import compile from 'wat-compiler'
import { cos, env, exp, fill, mod, sin } from '../src'

let memory
const make = (code: string) => {
  memory = new WebAssembly.Memory({
    initial: 16,
    maximum: 16,
    // shared: true,
  })

  const buffer = compile('(module ' + env() + code + ')')
  const mod = new WebAssembly.Module(buffer)
  const instance = new WebAssembly.Instance(mod, { env: { memory } }) as any
  return instance.exports as {
    fill(
      /** song position starting frame */
      startFrame: number,
      /** buffer length to fill */
      bufferLength: number,
      /** sample rate */
      sampleRate: number,
      /** input pointer */
      inputPtr: number,
      /** `f` routine params */
      ...args: number[]
    ): void
    f(x: number): number
  }
}

describe('fill()', () => {
  it('fills memory with the result of `f`', () => {
    const m = make(
      `;;wasm
      (func $f (result f32)
        (global.get $t)
      )
      ` + fill({ vars: [] })
    )

    m.fill(0, 10, 10, 0, 0, 0, 0)
    const floats = new Float32Array(memory.buffer, 0, 10)
    expect(floats).toMatchSnapshot()
  })

  it('can start at different frame', () => {
    const m = make(
      `;;wasm
      (func $f (result f32)
        (global.get $t)
      )
      ` + fill({ vars: [] })
    )

    m.fill(10, 10, 10, 0, 0, 0, 0)
    const floats = new Float32Array(memory.buffer, 0, 10)
    expect(floats).toMatchSnapshot()
  })

  it('accepts sample rate', () => {
    const m = make(
      `;;wasm
      (func $f (result f32)
        (global.get $t)
      )
      ` + fill({ vars: [] })
    )

    m.fill(0, 10, 20, 0, 0, 0, 0)
    const floats = new Float32Array(memory.buffer, 0, 10)
    expect(floats).toMatchSnapshot()
  })

  it('passes parameters to `f`', () => {
    const m = make(
      `;;wasm
      (func $f (param $a f32) (param $b f32) (param $c f32) (result f32)
        (f32.add (global.get $t)
        (f32.add (local.get $a)
        (f32.add (local.get $b) (local.get $c))))
      )
      ` + fill({ vars: ['a', 'b', 'c'] })
    )

    m.fill(0, 10, 10, 0, 1, 2, 3)
    const floats = new Float32Array(memory.buffer, 0, 10)
    expect(floats).toMatchSnapshot()
  })
})

describe('cos', () => {
  it('works', () => {
    const m = make(
      `;;wasm
      (func $f (export "f") (param $x f32) (result f32)
        (call $cos (local.get $x))
      )
      ` + cos
    )

    const x = [-2, -1, 0, 1, 2]
    expect(x.map(m.f)).toMatchSnapshot()
    expect(x.map(Math.cos)).toMatchSnapshot()
  })
})

describe('sin', () => {
  it('works', () => {
    const m = make(
      `;;wasm
      (func $f (export "f") (param $x f32) (result f32)
        (call $sin (local.get $x))
      )
      ` +
        cos + // sin depends on cos
        sin
    )

    const x = [-2, -1, 0, 1, 2]
    expect(x.map(m.f)).toMatchSnapshot()
    expect(x.map(Math.sin)).toMatchSnapshot()
  })
})

describe('exp', () => {
  it('works', () => {
    const m = make(
      `;;wasm
      (func $f (export "f") (param $x f32) (result f32)
        (call $exp (local.get $x))
      )
      ` + exp
    )

    const x = [-2, -1, 0, 1, 2]
    expect(x.map(m.f)).toMatchSnapshot()
    expect(x.map(Math.exp)).toMatchSnapshot()
  })
})

describe('mod', () => {
  it('works', () => {
    const m = make(
      `;;wasm
      (func $f (export "f") (param $x f32) (result f32)
        (call $mod (local.get $x) (f32.const 3.0))
      )
      ` + mod
    )

    const x = [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5]
    expect(x.map(m.f)).toMatchSnapshot()
    expect(x.map(x => x % 3)).toMatchSnapshot()
  })

  it('floats', () => {
    const m = make(
      `;;wasm
      (func $f (export "f") (param $x f32) (result f32)
        (call $mod (local.get $x) (f32.const 1.5))
      )
      ` + mod
    )

    const x = [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5]
    expect(x.map(m.f)).toMatchSnapshot()
    expect(x.map(x => x % 1.5)).toMatchSnapshot()
  })
})
