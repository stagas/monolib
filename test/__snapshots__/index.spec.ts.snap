// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cos works 1`] = `
Array [
  -0.4157480299472809,
  0.5404837131500244,
  1,
  0.5404837727546692,
  -0.41574785113334656,
]
`;

exports[`cos works 2`] = `
Array [
  -0.4161468365471424,
  0.5403023058681398,
  1,
  0.5403023058681398,
  -0.4161468365471424,
]
`;

exports[`exp works 1`] = `
Array [
  0.1353352814912796,
  0.3678794503211975,
  1,
  2.7182817459106445,
  7.389056205749512,
]
`;

exports[`exp works 2`] = `
Array [
  0.1353352832366127,
  0.36787944117144233,
  1,
  2.718281828459045,
  7.38905609893065,
]
`;

exports[`fill() accepts sample rate 1`] = `
Float32Array [
  0,
  0.05000000074505806,
  0.10000000149011612,
  0.15000000596046448,
  0.20000000298023224,
  0.25,
  0.30000001192092896,
  0.3499999940395355,
  0.4000000059604645,
  0.44999998807907104,
]
`;

exports[`fill() can start at different frame 1`] = `
Float32Array [
  1,
  1.100000023841858,
  1.2000000476837158,
  1.2999999523162842,
  1.399999976158142,
  1.5,
  1.600000023841858,
  1.7000000476837158,
  1.7999999523162842,
  1.899999976158142,
]
`;

exports[`fill() fills memory with the result of \`f\` 1`] = `
Float32Array [
  0,
  0.10000000149011612,
  0.20000000298023224,
  0.30000001192092896,
  0.4000000059604645,
  0.5,
  0.6000000238418579,
  0.699999988079071,
  0.800000011920929,
  0.8999999761581421,
]
`;

exports[`fill() passes parameters to \`f\` 1`] = `
Float32Array [
  6,
  6.099999904632568,
  6.199999809265137,
  6.300000190734863,
  6.400000095367432,
  6.5,
  6.599999904632568,
  6.699999809265137,
  6.800000190734863,
  6.900000095367432,
]
`;

exports[`mod floats 1`] = `
Array [
  -0.5,
  -1,
  -0,
  -0.5,
  -1,
  0,
  1,
  0.5,
  0,
  1,
  0.5,
]
`;

exports[`mod floats 2`] = `
Array [
  -0.5,
  -1,
  -0,
  -0.5,
  -1,
  0,
  1,
  0.5,
  0,
  1,
  0.5,
]
`;

exports[`mod works 1`] = `
Array [
  -2,
  -1,
  -0,
  -2,
  -1,
  0,
  1,
  2,
  0,
  1,
  2,
]
`;

exports[`mod works 2`] = `
Array [
  -2,
  -1,
  -0,
  -2,
  -1,
  0,
  1,
  2,
  0,
  1,
  2,
]
`;

exports[`sin works 1`] = `
Array [
  -0.9097958207130432,
  -0.8421677947044373,
  0,
  0.8421676158905029,
  0.909795880317688,
]
`;

exports[`sin works 2`] = `
Array [
  -0.9092974268256817,
  -0.8414709848078965,
  0,
  0.8414709848078965,
  0.9092974268256817,
]
`;
